examples/simplex.dx

```diff
@@ -0,0 +1,596 @@
+This is done using records to construct index sets.
+The benefit, in principle, is that this both annotates the code, and
+avoids indexing errors.
+
+'This implementation roughly follows the two-stage algorithm outlined in
+[these notes](http://web.mit.edu/15.053/www/AMP-Chapter-02.pdf)
+and [these notes](http://www.eng.uwaterloo.ca/~syde05/phase1.pdf).
+
+'This is a basic implementation, and doesn't allow equality constraints
+or bounded variables.
+
+
+import plot
+
+
+'# General utility functions.
+
+-- Hopefully these instances can one day be generated automatically for
+-- all record types.
+instance [Eq cons, Eq a] Eq {constraints: cons | objective: Unit | extra:a}
+  (==) = \r1 r2.
+    case r1 of
+      {|objective=r1|} ->
+        case r2 of
+          {|objective=r2|} -> True
+          {|extra=r2|} -> False
+          {|constraints=r2|} -> False
+      {|extra=r1|} ->
+        case r2 of
+          {|objective=r2|} -> False
+          {|extra=r2|} -> r1 == r2
+          {|constraints=r2|} -> False
+      {|constraints=r1|} ->
+        case r2 of
+          {|objective=r2|} -> False
+          {|extra=r2|} -> False
+          {|constraints=r2|} -> r1 == r2
+
+instance [Eq vars, Eq cons] Eq {variables:vars | slacks:cons | value: Unit}
+  (==) = \r1 r2.
+    case r1 of
+      {|variables=r1|} ->
+        case r2 of
+          {|variables=r2|} -> r1 == r2
+          {|slacks=r2|} -> False
+          {|value=r2|} -> False
+      {|slacks=r1|} ->
+        case r2 of
+          {|variables=r2|} -> False
+          {|slacks=r2|} -> r1 == r2
+          {|value=r2|} -> False
+      {|value=r1|} ->
+        case r2 of
+          {|variables=r2|} -> False
+          {|slacks=r2|} -> False
+          {|value=r2|} -> True
+
+
+instance [Eq a, Eq b] Eq {artificials: a | realvars: b}
+  (==) = \r1 r2.
+    case r1 of
+      {|artificials=r1|} ->
+        case r2 of
+          {|artificials=r2|} -> r1 == r2
+          {|realvars=r2|} -> False
+      {|realvars=r1|} ->
+        case r2 of
+          {|artificials=r2|} -> False
+          {|realvars=r2|} -> r1 == r2
+
+
+def argmax_suchthat [Ord o] (xs:n=>o) (cond:o->Bool): n =
+  -- Assumes that there's at least one element of xs
+  -- that satisfies the condition.
+  cmp = \x y. (not $ cond y) || (x > y && cond x)
+  argscan cmp xs
+
+def argmax_suchthat_table [Ord o] (xs:n=>o) (cond:n=>Bool): n =
+  -- This variant takes a table of pre-evaluated conditions.
+  cmp = \i j. (not $ cond.j) || (xs.i > xs.j && cond.i)
+  argscan cmp (for i. i)
+
+
+'### Simplex data types
+
+data SimplexResult n:Type =
+  Infeasible
+  Unbounded
+  Solution (n=>Float)
+
+data ConstraintType =
+  LessThan
+  GreaterThan
+  -- Todo: Add equality constraints
+
+data ObjectiveType =
+  Maximize
+  Minimize
+
+def Constraint (n:Type) : Type =
+  {coeffs:n=>Float & offset:Float & conType:ConstraintType}
+
+def ColIx (cons:Type) (vars:Type) : Type =
+  {variables:vars | slacks:cons | value: Unit}
```

Comment:
What does cons mean here?

Reply:
That's constraints here. I added comments for clarity.